examples/benchmark/datagen.py

```diff
@@ -0,0 +1,345 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+import abc
+import io
+import os
+import tarfile
+import time
+from multiprocessing import Queue
+from threading import Thread, Lock, Barrier
+from typing import Any, Callable, Iterator, TypeVar, NewType
+
+import boto3
+import click
+import numpy as np
+import prefixed as pr
+import yaml
+from PIL import Image
+
+
+class DataGenerator(abc.ABC):
+    @abc.abstractmethod
+    def create(self, idx_gen: Iterator[int]) -> Iterator[tuple[str, io.BytesIO]]:
+        pass
+
+
+class ImageGenerator(DataGenerator):
+    def __init__(self, width: int, height: int, img_format: str = "JPEG"):
+        self.width = width
+        self.height = height
+        self.img_format = img_format
+
+    def create(self, idx_iter: Iterator[int]) -> Iterator[tuple[str, io.BytesIO]]:
+        for index in idx_iter:
+            img = Image.fromarray(
+                np.random.randint(
+                    0, high=256, size=(self.height, self.width, 3), dtype=np.uint8
+                )
+            )
+            byte_buf = io.BytesIO()
+            img.save(byte_buf, format=self.img_format)
+            byte_buf.seek(0)
+
+            yield f"image_{index}.{self.img_format}", byte_buf
+
+
+class ThreadSafeCounter:
+    def __init__(self) -> None:
+        self.val = 0
+        self.lock = Lock()
+
+    def get_and_inc(self):
+        with self.lock:
+            self.val += 1
+            return self.val
+
+
+class Utils:
+    T = TypeVar("T")
+
+    @staticmethod
+    def batcher(
+        items: Iterator[T], size_extractor: Callable[[T], int], size_threshold: float
+    ) -> Iterator[list[T]]:
```

Comment:
Similarly, we can either use `List[]` or do a future annotations import for py3.8 support